### ISSUE TITLE
Optimize proxy to pod|service mapping computation and usage

### ISSUE THREAD

**Issue by shashankram** [openservicemesh/osm#4249]
**Please describe the Improvement and/or Feature Request**

The current code is pretty suboptimal because:
1. We compute proxy->service and proxy->pod mappings repeatedly on every update to proxies, even when there has been no change to those mappings (pod or service updates in the cluster).
2. Repeated cache access on thread-safe structures results in contention, which could be avoided.

I was thinking we could do the following:
When a service/pod event is received, don't generate a proxy broadcast from the broker, but compute proxy-service && pod-service mappings, and trigger updates to relevant proxies. As a start, we can broadcast the update and optimize on filtering proxies as a next step

With the above optimization, there would be no need to look up all pods/services in the mesh for proxy config generation when there are no pod/service changes in the cluster, significantly resulting in performance in scale environments.

**Scope (please mark with X where applicable)**
- Performance          [ ]
- Envoy Control Plane        [X]

**Possible use cases**
Performance optimization for scale environments.

**Comment by nojnhuh** [comment]
tl;dr I'm not 100% convinced (the way I've implemented) this results in any significant performance gains.

I have a working version of these changes on a branch here: https://github.com/openservicemesh/osm/compare/main...nojnhuh:svcreg. (In case I update that branch, it's currently pointing to https://github.com/openservicemesh/osm/commit/f2643c6228be113e35acf897f4ccfdf120e1ab63).

To summarize how this works, a new background goroutine is spun up to listen for events related to Kubernetes Endpoints resources. When an event comes through, a local cache mapping proxies to services is updated. Queries to this cache when XDS responses are generated are then simply `sync.RWMutex`-protected map lookups. For context, currently when any XDS handler needs to know which services are associated with a proxy, the `k8s.Controller`'s local caches are traversed to rebuild that mapping from scratch each time.

My changes do not address triggering updates for individual proxies based on each event and instead update all proxies for any Endpoints event seen. I hoped that we would only need to update the proxies associated with a service when its corresponding endpoints were updated, but I did run into one case where we would also need to update any proxy allowed to communicate with a service whose endpoints were updated, which would require much more bookkeeping. The issue I saw was, roughly:
- mysql Endpoints created with no ports listed
- bookwarehouse proxy connects and gets initial config. mysql cluster is not programmed because the target port on the corresponding MeshService is 0.
- mysql Endpoints updated with mysql Pod, only an update for the mysql proxy is triggered
- No broadcasts are triggered, so bookwarehouse config remains stale (no mysql cluster).

To test how these changes compare with the existing behavior, I ran both the main branch and my changes through two different scenarios. All tests were run on an AKS cluster with 6 Standard\_DS2\_v2 nodes in the context of the bookstore demo with each service scaled up to 10 replicas. The main branch for my tests was at 8dccabbf98d4187d7072b2145b1e280d45303f6d. Each test installs the demo, runs some operation repeatedly, runs a CPU profile for 15 minutes, takes a memory heap profile, then takes a Prometheus snapshot.

The first scenario constantly churns Endpoints as fast as possible:
```bash
while true; do kubectl delete ep -n bookstore --all; done
```
This is essentially the worst-case scenario for the async mapper where it does as much work as possible.

The second scenario constantly updates a label on a traffic target:
```bash
while true; do
  kubectl label traffictarget -n bookstore --overwrite bookbuyer-access-bookstore-v1 churn=$(date +%S)
  sleep 2 # osm-controller coalesces proxy broadcasts triggered less than 2 seconds apart
done
```
The particular operation isn't important here, just that it triggers a proxy update broadcast and that it doesn't add/update/delete any Endpoints. This is the best-case scenario for the async mapper, where it will build the cache once and never need to update it.

The results are summarized below:

scenario | log level | branch | Total CPU% | % of total CPU% spent mapping proxies | Memory heap
-|-|-|-|-|-
endpoint-updates | disabled | main | 2.40% | 2.31% | 128.24MB
endpoint-updates | disabled | svcreg | 3.16% | 0.21% | 109.83MB
endpoint-updates | trace | main | 3.93% | 2.63% | 87.83MB
endpoint-updates | trace | svcreg | 5.27% | 6.96% | 116.98MB
traffictarget-label | disabled | main | 4.56% | 4.97% | 25.34MB
traffictarget-label | disabled | svcreg | 6.16% | 0.036% | 25.8MB
traffictarget-label | trace | main | 9.29% | 5.59% | 24.12MB
traffictarget-label | trace | svcreg | 8.78% | 0.90% | 22.08MB

Some notes on the above table I find surprising:
- The "% of total CPU% spent mapping proxies" for the endpoint-updates scenario on svcreg branch is much higher for trace logging than disabled logging. In the profile, I can see that the time spent logging does not come close to making up that whole difference.
- For the traffictarget-label tests with disabled logging, the "Total CPU%" for the svcreg branch is slightly higher than the main branch despite the "% of total CPU spent mapping proxies" being almost 0.

The full data and extra scripts I used:  [mapping.tar.gz](https://github.com/openservicemesh/osm/files/7706253/mapping.tar.gz)
In that tarball, each of the `endpoint-updates` and `traffictarget-label` directories contain the following:
  - `snapshots/`: Prometheus snapshots that can be imported to another Prometheus instance (https://suraj.io/post/how-to-backup-and-restore-prometheus/)
  - `*.pb.gz`: pprof profiling output named for the image tag, log level, and type (cpu/heap). Inspect with `go tool pprof`.

<details>
<summary>Full steps to gather the same data:</summary>

1. Build and push OSM images off of the `main` branch. I tagged mine `latest`.
2. Build and push OSM images from my branch linked above. I tagged these `svcreg`.
3. Select a scenario to run at the bottom of scenario.yaml from the above tarball.
4. Run profile.sh from the above tarball. (This will take about 75-90 minutes)
    - I didn't try to make this portable at all, so you might need tweaks to get this to run.
5. `snapshots/` and `*.pb.gz` end up in `~/Desktop`. Make sure to move these before running more scenarios.
</details>

Overall, considering the profiling results are inconclusive at best, I'm not sure it makes sense to add these changes, at least the way I've implemented them.
